### ISSUE TITLE
Unable to update some split packages

### ISSUE THREAD

**Issue by branoholy** [rmarquis/pacaur#453]
At first I have to say you done a great work. I love pacaur the most from all AUR helpers.

Unfortunately, I have a problem with updating of split packages.

The bug occurred while updating my own split package [regilo](https://aur.archlinux.org/pkgbase/regilo/) containing packages [regilo](https://aur.archlinux.org/packages/regilo/) and [regilo-lib](https://aur.archlinux.org/packages/regilo-lib/). The package `regilo` depends on `regilo-lib` which is similar to `boost` and `boost-libs` so I do not think it is a problem with my PKGBUILD. It seems like that the dependencies between split packages is the problem. 

This error message has been printed:

```text
:: Installing regilo,regilo-lib package(s)...
loading packages...
resolving dependencies...
looking for conflicting packages...
error: failed to prepare transaction (could not satisfy dependencies)
:: regilo: installing regilo-lib (2.3.1-1) breaks dependency 'regilo-lib=2.3'
```

Run the following commands to reproduce the bug:

```bash
wget "https://aur.archlinux.org/cgit/aur.git/plain/PKGBUILD?h=regilo&id=b86464f4d8d46086bcca724ac25a4ac22e1a80dc" -O PKGBUILD
makepkg -sri
pacaur -Syu
```

I think I found the reason in your source code. Line [#958](https://github.com/rmarquis/pacaur/blob/master/pacaur#L958) tries to install `regilo-lib` with `--asdeps` (because `regilo` depends on `regilo-lib`) and the next line tries to install `regilo` normally (explicitly). The problem is that the packages are installed separately, by two `pacman` calls.

The same problem occurs when I try to install only `regilo-lib` from a file:

```text
sudo pacman -U regilo-lib-2.3.1-1-x86_64.pkg.tar.xz
```

But it **does not** occur when I install both packages at once:

```text
sudo pacman -U regilo-2.3.1-1-x86_64.pkg.tar.xz regilo-lib-2.3.1-1-x86_64.pkg.tar.xz
```

To fix this bug, it would be great if all packages were installed explicitly and the dependencies would be marked after installation with:

```text
sudo pacman -D --asdeps regilo-lib
```

Note: I know I could avoid this problem by setting the dependency from `regilo-lib=$pkgver` to `regilo-lib>=$pkgver` or `regilo-lib` (without version).

**Comment by rmarquis** [comment]
Thanks for the detailed bug report. I see your note at the end of report, but I still don't understand why you need to add this versioned dependency. This is entirely unnecessary:

* All subpackages of a split PKGBUILD have the same version,
* When the PKGBUILD is updated, both split packages will be seen as being outdated,
* pacaur will correctly handles dependency across split PKGBUILD and both will be updated.

In a nutshell, you're creating an issue here without reason. Versioned dependencies rarely make sense in a rolling release distribution, and here it doesn't seem to me that it does.

If you think it does indeed make sense, please add additional information. Otherwise, you should simply fix your PKGBUILD.

**Comment by branoholy** [comment]
You are right that the versioned dependency is not necessary in my PKGBUILD (that's why I wrote the note) or neither I do not see the reason. I made my PKGBUILD according to the boost [PKGBUILD](https://projects.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/boost) (the structure is similar to mine - separated headers and library) and they use versioned dependency as well.

But this bug report is not about my PKGBUILD. Another software in AUR could have the same problem right now or in the future. The only reason why I'm reporting this bug is because I found a **valid** configuration of PKGBUILD that **is not** processed correctly with pacaur. And I think it should be processed correctly because it is valid, no matter how rare it is.

**Comment by rmarquis** [comment]
Well, I'd argue that the versioned dependency isn't needed. Theoretically, you could add versioned dependencies everywhere, but they simply don't make sense in a rolling distribution where partial upgrades aren't supported.

In fact, the only situation I see this could be useful is in case of partial upgrade (having both split package installed, but then forcing reinstall of only one of these after a `pacman -Sy`). I'd still encourage you to get rid of that versioned string since it has no purpose - your PKGBUILD would be cleaner without it.

However, it is true that there have been regressions with split packages installation since the pacman 5.0 release and the removal of the `--pkg` option in makepkg (see #364). I fixed a few already, and I'll see how I can refactor better these pacman -U calls with the `--asdep`/`--asexp` options. In particular, I should be careful not to unnecessarily duplicate these --asdep calls, as it is currently done.

**Comment by rmarquis** [comment]
Code refactored in 32f2100fa53bfd370c16189b79a599bf2a8009f5. Your issue should now be solved too, let me know if that works as expected.

**Comment by branoholy** [comment]
Thanks for a quick fix, I appreciate it. I can confirm that my issue is solved and the split package was updated correctly.

However, the same problem could occur if the versioned dependency is set between two package bases. It would be necessary to run only one pacman call at all (not for every package base). Just wanting to let you know.

Anyway, thanks for the fix again.

**Comment by rmarquis** [comment]
> However, the same problem could occur if the versioned dependency is set between two package bases. It would be necessary to run only one pacman call at all (not for every package base). Just wanting to let you know.

I believe this should be covered already. Let me know if you are aware of an example.

Running pacman only once to install all packages at once is simply not possible, since some packages require their dependencies to be installed first before being built.

**Comment by branoholy** [comment]
You're right, it would be impossible to run only a single pacman call because of the `makedepends`. It might be possible to run one call for the build dependencies and the second one for the other packages.

Anyway, I have no problem with the current version. Thank you.

**Comment by rmarquis** [comment]
> It might be possible to run one call for the build dependencies and the second one for the other packages.

No, it's unfortunately not. Some dependencies are actually required to build *and* run the program, in which case they are only listed in the `depends` array.
